File: AppBox/ViewController/MenuHandler/NSApplication+MenuHandler.m
//
//  NSApplication+MenuHandler.m
//  AppBox
//
//  Created by Vineet Choudhary on 02/01/17.
//  Copyright © 2017 Developer Insider. All rights reserved.
//

#import "NSApplication+MenuHandler.h"

@implementation NSApplication (MenuHandler)


//MARK: - AppBox
- (IBAction)checkForUpdateTapped:(NSMenuItem *)sender {
    [UpdateHandler isNewVersionAvailableCompletion:^(bool available, NSURL *url) {
        if (available){
            [UpdateHandler showUpdateAlertWithUpdateURL:url];
        }else{
            [UpdateHandler showAlreadyUptoDateAlert];
        }
    }];
}

//MARK: - File
- (IBAction)viewLogFileTapped:(NSMenuItem *)sender {
	[[AppDelegate appDelegate] openLatestLogFile];
}

- (IBAction)preferencesTapped:(NSMenuItem *)sender {
    [PreferencesTabViewController presentPreferences];
}

//MARK: - Accounts
- (void)updateAccountsMenu{
    [self updateDropboxButton];
	
	//get user account details
	[[[DBClientsManager authorizedClient].usersRoutes getCurrentAccount] setResponseBlock:^(DBUSERSFullAccount * _Nullable result, DBNilObject * _Nullable routeError, DBRequestError * _Nullable networkError) {
		if (result) {
			@try {
				[UserData setLoggedInUserEmail:result.email];
				[UserData setLoggedInUserDisplayName:result.name.displayName];
				[self updateDropboxButton];
			} @catch (NSException *exception) {
				[EventTracker logExceptionEvent:exception];
			}
		}
	}];
    
    //get spaces usage
    [[[DBClientsManager authorizedClient].usersRoutes getSpaceUsage]
     setResponseBlock:^(DBUSERSSpaceUsage * _Nullable response, DBNilObject * _Nullable routeError, DBRequestError * _Nullable error) {
         if (response){
             @try {
                 NSNumber *usage = @(response.used.longValue / abBytesToMB);
                 NSNumber *allocated = @(response.allocation.individual.allocated.longValue / abBytesToMB);
                 
                 //save space usage in user default
                 [UserData setDropboxUsedSpace:usage];
                 [UserData setDropboxAvailableSpace:allocated];
                 [self updateDropboxButton];
                 
                 //log space usage
				 DDLogInfo(@"DropBox Used Space : %@MB", usage);
				 DDLogInfo(@"DropBox Available Space : %@MB", allocated);
                 
                 //check if dopbox running out of space
                 if ((allocated.integerValue - usage.integerValue) < abDropboxOutOfSpaceWarningSize){
                     [Common showAlertWithTitle:@"Warning" andMessage:[NSString stringWithFormat:@"You're running out of Dropbox space\n\n %@MB of %@MB used.", usage, allocated]];
                 }
             } @catch (NSException *exception) {
                 [EventTracker logExceptionEvent:exception];
             }
         }
     }];
}

- (void)updateDropboxButton{
	//Dropbox account menu
	NSString *email = [UserData loggedInUserEmail];
	
	NSMenuItem *dropboxAccountButton = [[AppDelegate appDelegate] dropboxAccountButton];
	NSString *dropboxAccountButtonTitle = [NSString stringWithFormat:@"Email: %@", email];
	[dropboxAccountButton setHidden:email.isEmpty];
	[dropboxAccountButton setTitle:dropboxAccountButtonTitle];
	
	//Dropbox account name menu
	NSString *displayName = [UserData loggedInUserDisplayName];
	NSMenuItem *dropboxNameButton = [[AppDelegate appDelegate] dropboxNameButton];
	NSString *dropboxNameButtonTitle = [NSString stringWithFormat:@"Name: %@", displayName];
	[dropboxNameButton setHidden:displayName.isEmpty];
	[dropboxNameButton setTitle:dropboxNameButtonTitle];
	
	//Dropbox space menu
    NSNumber *used = [UserData dropboxUsedSpace];
    NSNumber *available = [UserData dropboxAvailableSpace];
	
	NSMenuItem *dropboxSpaceButton = [[AppDelegate appDelegate] dropboxSpaceButton];
	NSString *dropboxSpaceButtonTitle = [NSString stringWithFormat:@"Usage: %@MB of %@MB used", used, available];
    [dropboxSpaceButton setTitle:dropboxSpaceButtonTitle];
}

- (IBAction)logoutDropBoxTapped:(NSMenuItem *)sender {
    NSAlert *alert = [[NSAlert alloc] init];
    [alert setMessageText: @"Are you sure?"];
    [alert setInformativeText:@"Do you want to logout current dropbox account?"];
	[alert setAlertStyle:NSAlertStyleInformational];
    [alert addButtonWithTitle:@"Yes"];
    [alert addButtonWithTitle:@"No"];
    if ([alert runModal] == NSAlertFirstButtonReturn){
        [[NSNotificationCenter defaultCenter] postNotificationName:abDropBoxLoggedOutNotification object:sender];
        [sender setEnabled:NO];
    }
}

- (IBAction)dropboxSpaceTapped:(NSMenuItem *)sender {
    [self updateAccountsMenu];
}

//MARK: - Help
- (IBAction)helpButtonTapped:(NSMenuItem *)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:abDocumentationURL]];
    [EventTracker logEventWithType:LogEventTypeExternalLinkHelp];
}

- (IBAction)latestNewsTapped:(NSMenuItem *)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:abTwitterURL]];
    [EventTracker logEventWithType:LogEventTypeExternalLinkTwitter];
}

- (IBAction)releaseNotesTapped:(NSMenuItem *)sender {
    [EventTracker logEventWithType:LogEventTypeExternalLinkReleaseNote];
    NSString *versionString = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"CFBundleShortVersionString"];
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:[NSString stringWithFormat:@"%@%@",abGitHubReleaseBaseURL,versionString]]];
}

- (IBAction)showDashboardTapped:(NSMenuItem *)sender {
}

- (IBAction)showLocalServerTapped:(NSMenuItem *)sender {
}

- (IBAction)licenseTapped:(NSMenuItem *)sender {
    [EventTracker logEventWithType:LogEventTypeExternalLinkLicense];
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:abLicenseURL]];
    
}
@end

File: AppBox/ViewController/MenuHandler/NSApplication+MenuHandler.swift
//
//  NSApplication+MenuHandler.swift
//  AppBox

import AppKit
import os
import AppBoxCore

extension NSApplication {

    private static let log = Logger(subsystem: "com.developerinsider.AppBox", category: "MenuHandler")

    // MARK: - AppBox

    func checkForUpdateTapped(_ sender: NSMenuItem?) {
        UpdateHandler.isNewVersionAvailable { available, url in
            if available {
                if url == nil && UpdateHandler.isInstalledViaHomebrew() {
                    UpdateHandler.showHomebrewUpdateAlert()
                } else {
                    UpdateHandler.showUpdateAlert(withUpdateURL: url)
                }
            } else {
                UpdateHandler.showAlreadyUptoDateAlert()
            }
        }
    }


    // MARK: - Dropbox account (refresh identity + space usage into UserData; surfaced in the Accounts pane)

    public func updateAccountsMenu() {
        DropboxTransport.shared.currentAccount { email, displayName, _ in
            DispatchQueue.main.async {
                if let email {
                    UserData.setLoggedInUserEmail(email)
                    UserData.setLoggedInUserDisplayName(displayName)
                }
            }
        }

        DropboxTransport.shared.spaceUsage { usedMB, allocatedMB, error in
            DispatchQueue.main.async {
                guard error == nil else { return }
                UserData.setDropboxUsedSpace(NSNumber(value: usedMB))
                UserData.setDropboxAvailableSpace(NSNumber(value: allocatedMB))
                Self.log.info("DropBox Used Space : \(usedMB)MB")
                Self.log.info("DropBox Available Space : \(allocatedMB)MB")

                if (allocatedMB - usedMB) < 150 {
                    _ = Common.showAlert(withTitle: "Warning",
                                         andMessage: "You're running out of Dropbox space\n\n \(usedMB)MB of \(allocatedMB)MB used.")
                }
            }
        }
    }

    // MARK: - Help

    func helpButtonTapped(_ sender: NSMenuItem?) {
        if let url = URL(string: "https://docs.getappbox.com") { NSWorkspace.shared.open(url) }
    }

    func releaseNotesTapped(_ sender: NSMenuItem?) {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
        if let url = URL(string: "https://github.com/getappbox/AppBox-iOSAppsWirelessInstallation/releases/tag/\(version)") {
            NSWorkspace.shared.open(url)
        }
    }

    func licenseTapped(_ sender: NSMenuItem?) {
        if let url = URL(string: "https://github.com/getappbox/AppBox-iOSAppsWirelessInstallation/blob/master/LICENSE.md") {
            NSWorkspace.shared.open(url)
        }
    }
}
